---
title: "Workshop 2/1: Gene set enrichment analysis"
subtitle: "Functional analyses in transcriptomics"
author: "January Weiner"
date: "`r Sys.Date()`"
output: 
    html_document:
      toc: true
      toc_float: true
      code_folding: show
outputdupa:
  xaringan::moon_reader:
    self-contained: true
    css: ["default", "files/cubi-fonts.css", "files/style.css" ]
    lib_dir: libs
    nature:
      highlightStyle: github
      countIncrementalSlides: false
      ratio: "16:9"
      beforeInit: "files/macros.js"
toc: no
---

```{r,echo=FALSE}
## Set default options for the knitr RMD processing
knitr::opts_chunk$set(echo=TRUE,warning=FALSE,message=FALSE,fig.width=5,fig.height=5,cache=FALSE,autodep=TRUE, results="hide")
library(pander)
library(knitr)
library(kableExtra)
library(ggthemes)
library(ggplot2)
theme_set(theme_tufte())
options(width=100)
```

```{r echo=FALSE,include=FALSE,eval=TRUE}
options(crayon.enabled = TRUE)
options(crayon.colors = 256)
knitr::knit_hooks$set(output = function(x, options){
  paste0(
    '<pre class="r-output"><code>',
    fansi::sgr_to_html(x = htmltools::htmlEscape(x), warn = FALSE),
    '</code></pre>'
  )
})

## this is an ugly, ugly hack, but otherwise crayon does not LISTEN TO REASON!!!
num_colors <- function(forget=TRUE) 256
library(crayon)
assignInNamespace("num_colors", num_colors, pos="package:crayon")
```


# Preparation

## Plan for today

 * Download the required data and software
 * View analysis report from CUBI
 * Repeat the differential expression analysis on the full data set
 * Run a number of enrichment analyses


## Software required

Make sure that you have the following libraries installed:

```{r}
library(tidyverse)
library(ggplot2)
library(cowplot)
library(tmod)
library(colorDF)
library(DESeq2)
library(msigdbr)
```

If you lack some of the libraries, install them with `insall.packages`,
e.g.  `install.packages("tmod")`. To install `DESeq2` and `msigdbr`, run the following
code:

```{r eval=FALSE}
install.packages("BiocManager")
BiocManager::install("DESeq2")
BiocManager::install("msigdbr")
```



## Data required

You need to download the following data:

  * [counts RDS file](../../Datasets/counts.rds) (6 MB)
  * [covariate file](../../Datasets/covar.rds) 
  * [gene annotation file](../../Datasets/annot.rds) (0.5 MB)

Next:

 1. Create a new project in Rstudio
 2. Create a new Rmarkdown file for your analysis
 3. Optionally, open this Rmarkdown file
    ([weiner_functional_transcriptomics_workshop_2_1.rmd](weiner_functional_transcriptomics_workshop_2_1.rmd) – note that if you do so, you will 
    see all the solutions to the exercises. Don't do it if you don't want
    spoilers ;-)
 4. Load the data as follows:

```{r eval=FALSE}
counts <- readRDS("Datasets/counts.rds")
covar  <- readRDS("Datasets/covar.rds")
annot  <- readRDS("Datasets/annot.rds")
```

Note that the exact path depends where you have stored the data. For
example, I kept the structure from github, so in my case
I load the data as follows:

```{r}
counts <- readRDS("../../Datasets/counts.rds")
covar  <- readRDS("../../Datasets/covar.rds")
annot  <- readRDS("../../Datasets/annot.rds")
```

# 1. Differential expression analysis

## Inspecting a full report

While you are downloading / installing the data, let us take a closer look
at a [report generated by our automated pipeline](GSE156063_report.html).

## Basic analysis

First, let us look at the new covariate file:

```{r}
summary_colorDF(covar)
```

The data set is much more extensive compared to the previous one.


Next, we need to run the differential expression analysis. We have already
done that previously with the smaller data set, so here is

**Exercise 2/1. (15min)** Go to the 
[workshop 1/2 instructions](../workshop_1_2/weiner_functional_transcriptomics_workshop_1_2.html) and 
repeat the first steps (creation of the ds2 object, inspecting of the
covariate file and the results) with the full data set. Use `group`
variable from the `covar` data frame as your main covariate again.


**Solution.** Please try to solve this exercise by yourself. If you have
any issues, however, here is the solution (click on the "Code" button to
show the solution):

```{r class.source="fold-hide",cache=TRUE}
ds2 <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=covar,
                              design= ~ group)
ds2 <- DESeq(ds2)
```

The ds2 object has three coefficients:

```{r results="markdown"}
resultsNames(ds2)
```

Since we have used the formula `~ group`, there is an `Intercept`
coefficient (which is uninteresting, because it corresponds to the average
expression in the "no" group) and two others which correspond to the
differences between `other` and `no` and `SC2` and `no` groups.

To get the results for one of the coefficients, we need to use the
`results` function. We will also convert it to a regular data frame (the
DFrame object from DESeq2 sucks):

```{r}
res_other <- results(ds2, name="group_other_vs_no")
res_other <- as.data.frame(res_other)
res_SC2   <- results(ds2, name="group_SC2_vs_no")
res_SC2 <- as.data.frame(res_SC2)
```

**Exercise. (5min)** Merging the results to see whether log2FoldChanges are
corelated. Use the `merge` function with `by=0` to indicate that you are
merging by row names. You can use the `suffixes=c(".other", ".SC2")` to make
interpretation easier.

**Solution:**

```{r class.source="fold-hide"}
res_merged <- merge(res_other, res_SC2, by=0, 
                    suffixes = c(".other", ".SC2"))
```

We can now make a first log2 FC vs log2 FC plot and also calculate the
correlation coefficient:

```{r results="markdown"}
ggplot(res_merged, aes(x=log2FoldChange.other, y=log2FoldChange.SC2)) +
  geom_point() + geom_hline(yintercept=0) + geom_vline(xintercept=0)
cor.test(res_merged$log2FoldChange.other, res_merged$log2FoldChange.SC2)
```

**Exercise (5min)**: merge the results for both comparisons with the
annotation data frame. Use `by.x="PrimaryID"` and `by.y=0` – that indicates
that the column "PrimaryID" of the annotation data frame corresponds to the
rownames in the results data frame. Order the results by adjusted p-value.

**Solution.**

```{r class.source="fold-hide"}
res_other <- merge(annot, res_other, by.x="PrimaryID", by.y=0) 
res_other <- res_other[ order(res_other$padj), ]
## or, alternatively with tidyverse
res_other <- res_other %>% arrange(padj)
res_SC2 <- merge(annot, res_SC2, by.x="PrimaryID", by.y=0)
res_SC2 <- res_SC2 %>% arrange(padj)
```

## Plotting the results

To plot the results, we need normalized and logarithmized count data (we
did this during the previous workshop). In DESeq2, instead of using
log(counts per million) we can also use the variance stabilizing
transformation (vst) function to get a
normalized count matrix.

```{r}
norm_counts <- assay(vst(ds2))
```

To make our lifes slightly easier, we will now define a function that plots
a gene given a covariate, normalized counts and an ID:

```{r}
plot_gene <- function(id, expr, covariate) {

  df <- data.frame(Expression=expr[id, ], Covariate=covariate)
  ggplot(df, aes(x=Covariate, y=Expression)) + 
    geom_boxplot() +
    geom_jitter(width = .2)
}

plot_gene("ENSG00000185745", norm_counts, covar$group)
```

**Exercise. (5min)** Plot top three genes from both comparisons.


## Gene set enrichment analysis

### Simplest analysis with builtin gene sets

Here is the code to run the gene set enrichment using `tmod` and both the
hypergeometric test and the CERNO test:

```{r}
## define the "foreground" set
pval_thr <- 1e-3 # define the p value threshold
lfc_thr  <- 1    # define the LFC threshold
sel <- abs(res_other$log2FoldChange) > lfc_thr &
       res_other$padj < pval_thr

fg <- res_other$SYMBOL[sel]
bg <- res_other$SYMBOL

hg_res_other <- tmodHGtest(fg=fg, bg=bg)

## it is easier with CERNO

gg <- res_other$SYMBOL[ order(res_other$pvalue) ]
ce_res_other <- tmodCERNOtest(gg)
```

Note that we are ordering the genes by p-value and not by the adjusted
p-value. What difference does it make?

**Exercise (10min).** Run the same enrichments for SC2 vs no virus
comparison. Inspect the results.

**Solution.**

```{r class.source="fold-hide"}
## we will shorten the code a bit.
sel <- abs(res_SC2$log2FoldChange) > lfc_thr &
       res_SC2$padj < pval_thr

hg_res_sc2 <- tmodHGtest(fg=res_SC2$SYMBOL[sel], bg=res_SC2$SYMBOL)
ce_res_sc2 <- tmodCERNOtest(res_SC2$SYMBOL[order(res_SC2$pvalue)])
```


So far so good, but how do we know the enrichment is real? Can we check it
somehow?

### Inspecting the results


#### Overview: tmodPanelPlots

A tmodPanelPlot juxtaposes enrichment analysis results from different
comparisons. For this, we need a list.

```{r fig.width=10,fig.height=10}
res <- list(other=ce_res_other, SC2=ce_res_sc2)
tmodPanelPlot(res, filter.rows.pval = 1e-3)
```

However, this does not tell us whether genes in a gene set go up or down?
Unfortunately, we need to decide first what it means that a gene "goes up"
or "goes down", that is, indicate which genes are DEGs.

```{r fig.width=10,fig.height=10}
res_merged <- merge(res_other, res_SC2, 
  by=c("PrimaryID", "ENSEMBL", "SYMBOL") , suffixes=c(".other", ".SC2"))
lfcs <- res_merged[ , c("log2FoldChange.other", "log2FoldChange.SC2") ]
pvals <- res_merged[ , c("padj.other", "padj.SC2") ]
pie <- tmodDecideTests(res_merged$SYMBOL, lfc = lfcs, pval=pvals)
names(pie) <- names(res)
tmodPanelPlot(res, filter.rows.pval = 1e-3, pie=pie, grid="b")

```


#### Evidence plots (ROC curves)

One way of inspecting the results is to plot the ROC curve associated with
the given module. Here is an example:

```{r}
gg <- res_SC2$SYMBOL[ order(res_SC2$pvalue) ]
evidencePlot(gg, m="LI.M75", gene.labels=TRUE)
```

**Exercise (10min).** Compare the evidence plots for `res_SC2` and gene
sets (modules) `LI.M75`, `LI.M127` and `LI.M37.0`.


#### Inspecting the individual genes.

So what genes are exactly in the gene set `LI.M75`? Can we display their
expression somehow?

```{r}
mm <- getModuleMembers("LI.M75")[[1]]
print(mm)
```

We now can retrieve the genes in the `annot` data frame and plot them.
(for simplicity, we take only first 4)

```{r fig.height=10,fig.width=10}
sel <- res_SC2[ res_SC2$SYMBOL %in% mm, ][["PrimaryID"]]
plots <- lapply(sel[1:4], plot_gene, expr=norm_counts, covariate=covar$group)
plot_grid(plotlist=plots, ncol=2)
```

#### Calculating eigengenes

Eigengenes are a sort of "standardized" expression for a whole gene set. We
define them as the first component of the PCA for genes that belong in the
given gene set. We can plot the eigengenes to show how the whole gene set
behaves in a context.

```{r}
eig <- eigengene(norm_counts, annot$SYMBOL)
df <- data.frame(covar, eigengene=eig[ "LI.M75", ])
ggplot(df, aes(x=group, y=eigengene)) + 
  geom_boxplot() +
  geom_jitter(width=.3)
```

#### Upset plots

Many genes are present in more than one gene set. Therefore, a lot of the
results are actually redundant. One way of visualizing this are upset
plots.

```{r fig.width=10,fig.height=10}
upset(ce_res_sc2$ID[1:20])
```


### Using an another source of gene sets

In the following we will use the excellent collection of gene sets from the
MSigDB database, which is conveniently provided by the `msigdbr` package.
To use it with tmod, we need to convert it into a tmod object. However, we
don't want to use the whole data base (which is huge), and instead we will
focus on the gene ontology data sets.

```{r msigdbr,cache=TRUE}
library(msigdbr)
mdf <- msigdbr(species = "Homo sapiens") # takes a long time
go <- mdf[ mdf$gs_subcat == "GO:BP", ]
go_mset <- makeTmodFromDataFrame(go, feature_col = "gene_symbol",
                                 module_col = "gs_id",
                                 title_col = "gs_name")
```

We can now use the `go_mset` object in every tmod function, for example:

```{r results="markdown"}
gg <- res_SC2$SYMBOL[ order(res_SC2$pvalue) ]
tmodCERNOtest(gg, mset=go_mset)
```

Aaand we have a problem: what we see on top are gene sets which contain
huge numbers of genes – but which are scantly enriched.

```{r}
evidencePlot(gg, m="M3458", mset=go_mset)
```

So we either have to filter the results or, better, filter the gene sets.
First, we calculate the number of genes per gene set, then we select only
these gene sets that have at most 50 genes and at least 5 genes.

```{r go2,cache=TRUE}
N <- sapply(go_mset$MODULES2GENES, length)
sel <- names(N)[ N >= 5 & N <= 50 ]
go <- go[ go$gs_id %in% sel, ]
go_mset <- makeTmodFromDataFrame(go, feature_col = "gene_symbol",
                                 module_col = "gs_id",
                                 title_col = "gs_name")
```

This will result in a much more specific result list:

```{r results="markdown"}
tmodCERNOtest(gg, mset=go_mset)
evidencePlot(gg, m="M11929", mset=go_mset, gene.labels=TRUE)
```

# 2. Beyond a simple gene set enrichment analysis

Gene set enrichment analysis – esp. 2nd generation tools – can be used in
many different contextes. Here we will inspect three such applications.

## 2.1 Principal component analysis with tmod

Results of the principal component analysis (PCA) can be interpreted using
gene set enrichments. For every component, every gene gets an associated
weight (score, loading). The higher the absolute value of that association,
the more "important" that gene is for the given component. Let us take a
look.

However, we will select the 1000 genes with the top variance; this makes
the PCA "cleaner" as the genes with lower variance introduce a lot of noise.

```{r pca,cache=TRUE}
count_var <- apply(norm_counts, 1, var)
sel <- order(count_var, decreasing = TRUE)[1:1000]
norm_counts_sel <- norm_counts[sel, ]

## this little bit of R code gives as a matchin annotation data frame
annot_sel <- annot[ match(rownames(norm_counts_sel), 
                          annot$PrimaryID), ]
```

We now can proceed with the PCA analysis:

```{r fig.width=10}
pca <- prcomp(t(norm_counts_sel), scale=TRUE)
df <- cbind(covar, pca$x[, 1:5])
g1 <- ggplot(df, aes(x=PC1, y=PC2, color=group)) + geom_point()
g2 <- ggplot(df, aes(x=PC3, y=PC4, color=group)) + geom_point()
plot_grid(g1, g2)
```

We see that PC1 describes the difference between SC2 and the other two
groups, while PC3 corresponds to the difference between infected groups and
the healthy group. We now can ask what are the corresponding functional
differences. The loadings for the genes are in the `pca$rotation` matrix:

```{r results="markdown"}
gg <- annot_sel$SYMBOL[ order(-abs(pca$rotation[,3])) ]
tmodCERNOtest(gg)
```

And for PCA1:

```{r results="markdown"}
gg <- annot_sel$SYMBOL[ order(-abs(pca$rotation[,1])) ]
tmodCERNOtest(gg)
```

## Correlation analysis

We can attempt a functional annotation of a gene based on how it correlates
with other genes. Fore example, in the top 100 genes for `res_SC2` there is
the gene called `SLFN5`; the description says `schlafen family member 5`.
The entry in [Gene
Cards](https://www.genecards.org/cgi-bin/carddisp.pl?gene=SLFN5) is also
hardly illuminating. However, we might want to look at genes that are
correlated with SLFN5; they may shed light on its function.

```{r}
id <- "ENSG00000166750"
cor_mtx <- cor(t(norm_counts), norm_counts[id, ])
```

What is the top correlating gene?

```{r}
head(cor_mtx[ order(-abs(cor_mtx[,1])), ])
qplot(norm_counts[id, ], norm_counts[ "ENSG00000196141", ],
      color=covar$group)
```


```{r results="markdown"}
## make sure that the rownames are in the same order as annotation
all(rownames(cor_mtx) == annot$PrimaryID)

tmodCERNOtest(annot$SYMBOL[ order(-abs(cor_mtx[,1])) ])
```

No surprises here.

## Disco analysis

Let us now compare the concordant and discordant genes when comparing SC2
to the "other" group.

To calculate the disco score, we can use the `disco` package from Teresa
Domaszewska. However, it is quite easy to do it manually:


```{r}
res_merged <- merge(res_other, res_SC2, 
  by=c("PrimaryID", "ENSEMBL", "SYMBOL") , suffixes=c(".other", ".SC2"))

## we will use tidyverse for readability
res_merged <- res_merged %>% 
  mutate(disco = log2FoldChange.other * log2FoldChange.SC2 * -1 * 
         (log10(pvalue.other) + log10(pvalue.SC2)))
```

**Exercise. (5 min)** Plot the gene with the top and bottom disco score.

**Solution.**

```{r eval=FALSE,class.source="fold-hide"}
res_merged %>% arrange(-disco) %>% head()
res_merged %>% arrange(disco) %>% head()

plot_gene("ENSG00000169248", norm_counts, covar$group)
plot_gene("ENSG00000115008", norm_counts, covar$group)
```

We can also make a colored log2FC / log2FC plot. However, we need to
manipulate the disco values a bit: the concordant gene sets have high
positive outliers. Also, we need to set up a color scale.

```{r}
res_merged$disco2 <- ifelse(res_merged$disco > 50, 50, res_merged$disco)
ggplot(res_merged, aes(x=log2FoldChange.other, y=log2FoldChange.SC2,
                     color=disco2)) + geom_point(alpha=.5) +
  scale_color_gradient2(low="blue", mid="grey", high="red") +
  geom_hline(yintercept=0, color="grey") + geom_vline(xintercept=0, color="grey")
```


Finally, we can attempt to characterize functionally the discordant and
concordant genes.

```{r results="markdown"}
## discordant genes
tmodCERNOtest(res_merged$SYMBOL[ order(res_merged$disco) ])
## concordant genes
tmodCERNOtest(res_merged$SYMBOL[ order(-res_merged$disco) ])
```



# 3. Testing for interaction

One of the valid questions is whether transcriptomic responses in Sars Cov
2 are different in males than in females (given the different incidence of
Covid and severe Covid in the two sexes).

For this, we need to set up the DESeq2 model with a different equation that
will allow us to freely define the contrasts. But first we need to create a
variable that is a cross between the sex covariate and the group covariate.

```{r ds2_interaction,results="markdown", cache=TRUE}
covar$gender_group <- paste0(covar$gender, "_", covar$group)
ds2_int <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=covar,
                              design= ~ 0 + gender_group)
ds2_int<- DESeq(ds2_int)
resultsNames(ds2_int)
```

As you can see, we now have six coefficient in total, each one
corresponding to one level of the `gender_group` variable. We can now make
a more extensive test using *contrasts*. We specify the contrast using a
numeric vector of ${-1, 0, 1}$ which denotes whether a coefficient should
go to the contrast and which which sign.

For example, to compare expression in SC2 infected persons to healthy
controls in females, we can specify the contrast as follows:

```{r}
tmp <- results(ds2_int, contrast=c(-1, 0, 1, 0, 0, 0))
```

We would denote this contrast as `(F_SC2 - F_no)`.


The interaction contrast is the "difference of differences", i.e.
difference between the effect in females and the effect in males:

 $$(F_{\text{SC2}} - F_{\text{no}}) - (M_{\text{SC2}} - M_{\text{no}}) = $$

 
 $$F_{\text{SC2}} - F_{\text{no}} - (M_{\text{SC2}} + M_{\text{no}} = $$

Thus, we need to define contrast with $-1$ for `F_no` and `M_SC2` and $1$
for `F_SC2` and `M_no`:


```{r}
res_int_sc2 <- results(ds2_int, contrast=c(-1, 0, 1, 1, 0, -1))
res_int_sc2 <- as.data.frame(res_int_sc2)
res_int_sc2 <- merge(annot, res_int_sc2, by.x="PrimaryID", by.y=0)
```

(zeroes correspond to the `F_other` and `M_other` groups).

**Exercise. (15min)** Create the interaction contrast for `other` vs `no` / F
vs M. Run gene set enrichment analysis on both interaction contrasts with the CERNO test
and default gene sets. What do you see when you inspect the results?

**Solution.**

```{r class.source="fold-hide"}
res_int_other <- results(ds2_int, contrast=c(-1, 1, 0, 1, -1, 0))
res_int_other <- as.data.frame(res_int_other)
res_int_other <- merge(annot, res_int_other, by.x="PrimaryID", by.y=0)

ce_res_int_sc2 <- tmodCERNOtest(res_int_sc2$SYMBOL[
                                  order(res_int_sc2$pvalue) ])
ce_res_int_other <- tmodCERNOtest(res_int_other$SYMBOL[
                                  order(res_int_other$pvalue) ])
```


# Homework

**Option 1.** Prepare an Rmarkdown file which describes and performs the
basic steps of the analysis. It should include the following steps:

 * Differetial gene expression analysis of the SC2 data set with at least
   one comparison (e.g. SC2 vs no)
 * Gene set enrichment using tmod
 * Visualisation using tmodPanelPlot and evidencePlots 

You are free to use the code from the workshop Rmd files as your template.
*However*, you have to describe in *your own words* what you are doing and
why and what the results are.


**Option 2.** Provide any other analysis in any other literate programming
document system (LaTeX, Jupyter Notebook) which also includes differential
gene expression analysis and gene set enrichment analysis with tool(s) of
your choice.

**Option 3.** Send me the results of one of the optional exercises (below)

# Optional Exercises


For those of you who are bored:

**Comparing tmod to fgsea.** Create two additional data sets. One should
contain only 3 individuals per group, the other should include 20
individuals per group. Run the DE analysis for each of these data sets
(focusing on the comparison SC2 vs "other") and for the whole data set.

Next, run gene set enrichment analysis: once with tmod and once with fgsea.
Compare the results for each of the two smaller data sets with the results for the
full data set. Frankly, I have no idea what the result will be.

**Demonstrating the danger of Venn Diagrams.** First, create a data set
containing 40 SC2 patients and 40 healthy individuals. Create a
corresponding covariate data frame. Then, in the covariate data frame,
create two new groups: "A" and "B", randomly assigning the individuals to
one of these groups. Clearly, we cannot expect any systematic differences
between A and B. However, now we run the comparison between SC2 and healthy
individuals independently in each of the two groups (use contrasts for
that, creating first a dummy variable with levels `SC2_A`, `SC2_B`, `no_A`
and `no_B`). Plot a Venn diagram showing the comparison between results in
group A and B. Run a hypergeometric test analysis to find out whether genes
"specific" for data set A (i.e., significant only in A, but not in B
comparison) are enriched for a function.

