<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Lecture 4: Gene Set Enrichment Analysis</title>
    <meta charset="utf-8" />
    <meta name="author" content="January Weiner" />
    <meta name="date" content="2021-07-09" />
    <script src="libs/header-attrs-2.9/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
    <script src="libs/jquery-3.5.1/jquery.min.js"></script>
    <link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
    <script src="libs/datatables-binding-0.18/datatables.js"></script>
    <link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet" />
    <link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet" />
    <script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script>
    <link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
    <script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
    <link rel="stylesheet" href="files/cubi-fonts.css" type="text/css" />
    <link rel="stylesheet" href="files/style.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Lecture 4: Gene Set Enrichment Analysis
## Functional analyses in transcriptomics
### January Weiner
### 2021-07-09

---











&lt;!--
class:empty-slide,myinverse
background-image:url(images/arnolfini.jpg)
--&gt;


---

## Gene set enrichment analysis





The problem: how do you interpret a list of 1000 results?

---

<div id="htmlwidget-12624af77be37f690e4e" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-12624af77be37f690e4e">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100"],["RSAD2","OAS2","IFIT1","CMPK2","IFI44L","MX1","OAS3","HERC5","ZC3HAV1","IFIH1","IFI44","DDX58","IFIT2","IFI6","WARS1","IFIT5","PARP12","CXCL11","TNIP3","ISG15","IFIT3","PARP14","STAT2","CXCL10","EPSTI1","OASL","GBP4","SAMD9L","LAP3","LIPA","PML","DDX60","TNFSF10","BATF2","SERPING1","GZMB","SLAMF7","ZNFX1","APOL6","LAG3","LGALS9","USP18","GBP1","TRIM22","EIF2AK2","OAS1","STAT1","PARP9","HAVCR2","IDO1","HERC6","IFI35","CCL8","DTX3L","HES4","RNF213","CD69","LAMP3","IL2RA","LILRB4","IRF7","ACO1","PRF1","XAF1","TNS3","SOCS1","UBE2L6","TRIM38","IFITM3","ACOD1","TAP2","HAPLN3","IL4I1","SAMD9","APOL2","HAMP","MTURN","CTSL","MGAT1","CXCL9","FBXO6","HELZ2","DHX58","PATL1","NT5C3A","APOBEC3G","ATP10A","GCH1","ISG20","P2RX7","XRN1","TAP1","ETV7","HELB","TRIM21","APOL1","ZCCHC2","SP100","CD68","RASSF4"],["radical S-adenosyl methionine domain containing 2","2'-5'-oligoadenylate synthetase 2","interferon induced protein with tetratricopeptide repeats 1","cytidine/uridine monophosphate kinase 2","interferon induced protein 44 like","MX dynamin like GTPase 1","2'-5'-oligoadenylate synthetase 3","HECT and RLD domain containing E3 ubiquitin protein ligase 5","zinc finger CCCH-type containing, antiviral 1","interferon induced with helicase C domain 1","interferon induced protein 44","DExD/H-box helicase 58","interferon induced protein with tetratricopeptide repeats 2","interferon alpha inducible protein 6","tryptophanyl-tRNA synthetase 1","interferon induced protein with tetratricopeptide repeats 5","poly(ADP-ribose) polymerase family member 12","C-X-C motif chemokine ligand 11","TNFAIP3 interacting protein 3","ISG15 ubiquitin like modifier","interferon induced protein with tetratricopeptide repeats 3","poly(ADP-ribose) polymerase family member 14","signal transducer and activator of transcription 2","C-X-C motif chemokine ligand 10","epithelial stromal interaction 1","2'-5'-oligoadenylate synthetase like","guanylate binding protein 4","sterile alpha motif domain containing 9 like","leucine aminopeptidase 3","lipase A, lysosomal acid type","promyelocytic leukemia","DExD/H-box helicase 60","TNF superfamily member 10","basic leucine zipper ATF-like transcription factor 2","serpin family G member 1","granzyme B","SLAM family member 7","zinc finger NFX1-type containing 1","apolipoprotein L6","lymphocyte activating 3","galectin 9","ubiquitin specific peptidase 18","guanylate binding protein 1","tripartite motif containing 22","eukaryotic translation initiation factor 2 alpha kinase 2","2'-5'-oligoadenylate synthetase 1","signal transducer and activator of transcription 1","poly(ADP-ribose) polymerase family member 9","hepatitis A virus cellular receptor 2","indoleamine 2,3-dioxygenase 1","HECT and RLD domain containing E3 ubiquitin protein ligase family member 6","interferon induced protein 35","C-C motif chemokine ligand 8","deltex E3 ubiquitin ligase 3L","hes family bHLH transcription factor 4","ring finger protein 213","CD69 molecule","lysosomal associated membrane protein 3","interleukin 2 receptor subunit alpha","leukocyte immunoglobulin like receptor B4","interferon regulatory factor 7","aconitase 1","perforin 1","XIAP associated factor 1","tensin 3","suppressor of cytokine signaling 1","ubiquitin conjugating enzyme E2 L6","tripartite motif containing 38","interferon induced transmembrane protein 3","aconitate decarboxylase 1","transporter 2, ATP binding cassette subfamily B member","hyaluronan and proteoglycan link protein 3","interleukin 4 induced 1","sterile alpha motif domain containing 9","apolipoprotein L2","hepcidin antimicrobial peptide","maturin, neural progenitor differentiation regulator homolog","cathepsin L","alpha-1,3-mannosyl-glycoprotein 2-beta-N-acetylglucosaminyltransferase","C-X-C motif chemokine ligand 9","F-box protein 6","helicase with zinc finger 2","DExH-box helicase 58","PAT1 homolog 1, processing body mRNA decay factor","5'-nucleotidase, cytosolic IIIA","apolipoprotein B mRNA editing enzyme catalytic subunit 3G","ATPase phospholipid transporting 10A (putative)","GTP cyclohydrolase 1","interferon stimulated exonuclease gene 20","purinergic receptor P2X 7","5'-3' exoribonuclease 1","transporter 1, ATP binding cassette subfamily B member","ETS variant transcription factor 7","DNA helicase B","tripartite motif containing 21","apolipoprotein L1","zinc finger CCHC-type containing 2","SP100 nuclear antigen","CD68 molecule","Ras association domain family member 4"],[4.29980894178029,2.62607197671569,3.38482838176403,3.05323344681716,3.16785124765831,2.43794990702794,3.04822143740288,3.48848090760445,1.97463479486106,2.75819943873261,2.52458195081039,2.76857180295592,4.18759280902698,2.83997704464593,2.36489541303769,2.23937076446261,1.95889081723555,5.51518633903588,2.74860509232458,3.8128508208247,3.57665071216666,2.07785307105753,1.59922087322007,4.91730364016048,2.60647609841947,3.31151295373164,2.20571695794849,2.50475558176684,1.85467143784376,2.26943390663563,1.81606702188777,2.21654046366468,1.69983818893305,2.9638174654011,2.76414176308241,3.78637324056643,2.99646409690712,1.37975678765847,1.19659303644434,2.93661968032543,1.31469921374498,2.44453488664625,2.82793885865715,1.55549584117526,1.46646934732648,2.14912192314856,1.62790056698205,1.600598300277,2.02802821895529,2.75110418765495,1.86933541224611,1.80592809488897,6.69263686673274,1.3049626033272,2.16408805318111,1.8696328185906,3.17912777258872,1.91227197245234,2.50972164430418,2.25235968754991,2.25846359827086,1.64304632608884,3.07790533320178,1.52635258799544,2.05554635869814,2.58065124086593,1.75432640535088,1.2736151123336,2.1445516946784,6.11111628549307,1.83194264130962,2.35980519590627,2.97758037717603,1.80654665577582,1.85802083646642,2.46942494724421,-1.28833229028717,1.81821959479324,1.12543080913135,3.52079068803786,1.34466099578669,1.8603106935689,1.69201640638816,1.1235394905683,1.58153059315649,1.22145578903035,1.85662022353179,1.88727057925452,2.0784161316419,2.06538107270151,1.47497094334005,1.75505195364307,2.07838591330021,1.87373397675891,1.45045680292097,1.53299697112994,1.33420465656076,1.34050030607035,1.56211246394918,1.88141728924979],[1.00782604630107e-36,5.10619211207264e-33,4.38333055813318e-31,4.62144891258998e-31,4.76618071126657e-31,8.29390261060581e-31,9.25881091871409e-31,1.56360374896319e-30,4.68200341576065e-30,4.92138010228915e-28,1.05217621989292e-27,2.07270627171678e-27,2.55157535467568e-27,1.73197657456409e-25,1.73197657456409e-25,2.68178565677361e-25,4.74090637360152e-25,4.74090637360152e-25,1.55288995510735e-24,2.50781783898825e-24,3.77480399763908e-24,4.99759255922646e-24,6.20465137024061e-23,1.2337055935988e-22,1.4941315190416e-22,1.75460032865542e-22,1.87129679364508e-22,2.50761156851424e-22,6.69081385054334e-22,1.28114518003577e-21,1.28508856800242e-21,2.99216254208887e-21,5.53554566505081e-21,5.53554566505081e-21,8.54514507523826e-21,1.12410004767976e-20,2.42975761709681e-20,3.135736585768e-20,5.54025373692328e-20,7.6794890264823e-20,9.99864740342954e-20,9.99864740342954e-20,1.28540817126906e-19,1.99527608056575e-19,3.67528699695413e-19,5.63991939442455e-19,1.31395681058691e-18,1.55248764368058e-18,5.98133863966323e-18,5.98696652244198e-18,7.03118789041512e-18,8.11793047263556e-18,1.14868755038205e-17,2.66502561473368e-17,2.66502561473368e-17,3.16703359740015e-17,4.23959274862442e-17,4.32603063148384e-17,7.11337274532687e-17,1.62200286750344e-16,2.75348039057234e-16,2.81519611258622e-16,3.14677488482612e-16,3.38545269534951e-16,3.60321680650992e-16,3.76566962976674e-16,4.34470611583662e-16,4.53317709231466e-16,5.48932578508867e-16,6.13039593668145e-16,1.57122276859699e-15,1.76074381849542e-15,2.97500823509828e-15,3.36524718807141e-15,3.48671045400853e-15,4.97054656766858e-15,5.02011932195607e-15,5.18208464296797e-15,5.76204501225597e-15,6.44004230758062e-15,1.24278459751826e-14,1.24278459751826e-14,1.39941107835038e-14,1.52231406430103e-14,1.55159964216761e-14,2.08179049021292e-14,2.23123832615849e-14,2.81514355916996e-14,2.81514355916996e-14,2.83104351698371e-14,3.92534084244133e-14,5.02891918985985e-14,5.22322114697146e-14,5.22322114697146e-14,5.23974506461681e-14,5.32002480030987e-14,5.99166136007578e-14,6.42229029230787e-14,6.42229029230787e-14,7.21432313271957e-14]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>SYMBOL<\/th>\n      <th>GENENAME<\/th>\n      <th>log2FoldChange<\/th>\n      <th>padj<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render"],"jsHooks":[]}</script>

Total results: 1249 
(at `\(p &lt; 0.05\)` and `\(\log_2(FC) &gt; 1\)` )

---

## A few examples

![](weiner_gsea_04_files/figure-html/unnamed-chunk-5-1.svg)&lt;!-- --&gt;

???

OAS2: by creating oligomers of ATP, induces RNASEL

IFIT1: binds to PPP-RNA, acting as a sensor to viral RNA

IFI44L: negative feedback – lowers the interferon response

ZC3HAV1, aka ZAP (zinc finger antiviral protein), Viruses comprised a unique ZRE, such as MLVs, SINV, Ebola virus, HIV-1, and hepatitis B virus, are susceptible to ZC3HAV1

---

.mycenter[
Basic idea behind gene set enrichments:&lt;br&gt;
categorize genes
]


---
class:empty-slide,myinverse
background-image:url(images/desert2.jpg)

.mytop[
The number of grains of sand in this picture accurately represents the
number of approaches for functional analysis of gene expression.
]

---

## Gene ontology database

.pull-left[

 * Genes are assigned to GO categories
 * GO categories are assigned to each other
 * GO can be represented as a *directed acyclic graph*

]

.pull-right[
![](images/gorilla.png)
]

---

## KEGG pathways

KEGG – Kyoto Encyclopaedia of Genes and Genomes

---
class:empty-slide,myinverse
background-image:url(images/hsa04110.gse16873.png)

---


## Other gene sets

Different lists of genes can be obtained from various sources.

 * Coexpression data sets: groups of genes that have correlated expression
   profiles

--

 * Results from another differential gene expression analysis experiment:
   groups of genes significant in a comparison

--

 * Targets of a particular miRNA determined computationally


---

## MSigDB – a meta database of gene sets

.pull-left[
![:scale 60%](images/msigdb.png)
]

.pull-right[

(very) Important sets:

 * **Hallmark:** 50 manually selected gene sets
 * **C2:** Kegg, Reactome, Wikipathways, Biocarta
 * **C5:** Gene Ontology (GO)
]

---

## Group of genes, and then what?

**Question:** are genes from a gene set `\(S\)` more likely to be
differentially expressed than other genes?

--

 Gene set enrichment analysis:

 * 1st generation: hypergeometric test, Fisher's exact test
 * 2nd generation: other algorithms, often relying on ordering the list of
   genes
 * 3rd generation: pathway topology analysis


---

## Generation 1: 

First, define DEGs: differentially expressed genes (e.g.
`\(|\log_2(FC)| &gt; 1\)`, `\(q &lt; 0.001\)`)

Second, define which genes are in a given gene set (GS), and which are not.

.pull-left[

**No significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|5 | 95| 5%|
|non in GS|50 |1000|5%|
|Fraction|9%|9.5%||

]

--

.pull-right[


**Significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|50 | 60| 45%|
|non in GS|70 |1000|7%|
|Fraction|42%|6%||

]

How can we test it?

---

.pull-left[

**No significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|5 | 95| 5%|
|non in GS|50 |1000|5%|
|Fraction|9%|9.5%||

]

.pull-right[


**Significant enrichment:**

|    |DEGs |non-DEGs|Fraction|
|----|-----|--------|--------|
|in GS|50 | 60| 45%|
|non in GS|70 |1000|7%|
|Fraction|42%|6%||

]

--

.pull-left[
Results of `\(\chi^2\)` test:

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:right;"&gt; statistic &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; p.value &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; parameter &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; method &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Pearson's Chi-squared test with Yates' continuity correction &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

]

--

.pull-right[


Results of `\(\chi^2\)` test:


&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:right;"&gt; statistic &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; p.value &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; parameter &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; method &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 161.097 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 0 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Pearson's Chi-squared test with Yates' continuity correction &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

]

---

## Effect size in ORA

Given gene set `\(S\)`, we can define enrichment `\(E\)` as following odds ratio:

 `$$E =  \frac{\frac{b}{n}}{\frac{B}{N}} = \frac{b}{B} \cdot \frac{N}{n}$$`

where

 * `\(b\)` – number of DEGs in `\(S\)`
 * `\(n\)` - number of DEGs
 * `\(B\)` - number of genes in `\(S\)`
 * `\(N\)` – number of genes

(actually, we should use `\(B-b\)` and `\(N-n\)` for odds ratio)

---

## Example results 

For DEG defined with `\(q &lt; 0.001\)`, `\(|\log_2(FC)| &gt; 1\)`

<div id="htmlwidget-d6efb2190b6896f25649" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-d6efb2190b6896f25649">{"x":{"filter":"none","data":[["LI.M7.1","LI.M7.0","LI.M7.3","LI.M127","LI.M7.2","LI.M75","LI.M165","LI.M13","LI.M150","LI.M111.1","LI.M68","LI.M27.1","LI.M86.1","LI.M67","LI.M5.0","LI.M27.0","LI.M29","LI.M38","LI.M43.1","LI.M35.1","LI.M45","LI.M86.0","LI.M66","LI.M5.1","LI.M11.0","LI.M121","LI.M35.0","LI.M93","LI.M19","LI.M64","LI.S0","LI.M14","LI.M36","LI.M4.3","LI.M111.0","LI.S11","LI.M43.0","LI.M126","LI.M4.0","LI.M12","LI.M89.1","LI.M115","LI.M112.0","LI.M18","LI.M119","LI.M52","LI.M20","LI.M24","LI.S10","LI.M59","LI.M61.2","LI.M65","LI.M112.1","LI.M95.0","LI.M117","LI.M44","LI.M91","LI.M62.0","LI.M37.0","LI.M31","LI.M61.0","LI.M53","LI.M82","LI.M157","LI.M164"],["LI.M7.1","LI.M7.0","LI.M7.3","LI.M127","LI.M7.2","LI.M75","LI.M165","LI.M13","LI.M150","LI.M111.1","LI.M68","LI.M27.1","LI.M86.1","LI.M67","LI.M5.0","LI.M27.0","LI.M29","LI.M38","LI.M43.1","LI.M35.1","LI.M45","LI.M86.0","LI.M66","LI.M5.1","LI.M11.0","LI.M121","LI.M35.0","LI.M93","LI.M19","LI.M64","LI.S0","LI.M14","LI.M36","LI.M4.3","LI.M111.0","LI.S11","LI.M43.0","LI.M126","LI.M4.0","LI.M12","LI.M89.1","LI.M115","LI.M112.0","LI.M18","LI.M119","LI.M52","LI.M20","LI.M24","LI.S10","LI.M59","LI.M61.2","LI.M65","LI.M112.1","LI.M95.0","LI.M117","LI.M44","LI.M91","LI.M62.0","LI.M37.0","LI.M31","LI.M61.0","LI.M53","LI.M82","LI.M157","LI.M164"],["T cell activation (I)","enriched in T cells (I)","T cell activation (II)","type I interferon response","enriched in NK cells (I)","antiviral IFN signature","enriched in activated dendritic cells (II)","innate activation by cytosolic DNA sensing","innate antiviral response","viral sensing &amp; immunity; IRF2 targets network (II)","RIG-1 like receptor signaling","chemokine cluster (II)","proinflammatory dendritic cell, myeloid cell response","activated dendritic cells","regulation of antigen presentation and immune response","chemokine cluster (I)","proinflammatory cytokines and chemokines","chemokines and receptors","myeloid, dendritic cell activation via NFkB (II)","signaling in T cells (II)","leukocyte activation and migration","chemokines and inflammatory molecules in myeloid cells","TBA","T cell activation and signaling","enriched in monocytes (II)","TBA","signaling in T cells (I)","TBA","T cell differentiation (Th2)","enriched in activated dendritic cells/monocytes","T cell surface signature","T cell differentiation","T cell surface, activation","myeloid cell enriched receptors and transporters","viral sensing &amp; immunity; IRF2 targets network (I)","Activated (LPS) dendritic cell surface signature","myeloid, dendritic cell activation via NFkB (I)","double positive thymocytes","cell cycle and transcription","CD28 costimulation","putative targets of PAX3","cytokines - recepters cluster","complement activation (I)","T cell differentiation via ITK and PKC","enriched in activated dendritic cells (I)","T cell activation (IV)","AP-1 transcription factor network","cell activation (IL15, IL23, TNF)","Resting dendritic cell surface signature","CCR1, 7 and cell signaling","enriched in NK cells (receptor activation)","IL2, IL7, TCR network","complement activation (II)","enriched in antigen presentation (II)","cell adhesion (GO)","T cell signaling and costimulation","adhesion and migration, chemotaxis","T &amp; B cell development, activation","immune activation - generic cluster","cell cycle and growth arrest","enriched in NK cells (II)","inflammasome receptors and signaling","signal transduction, plasma membrane","enriched in NK cells (III)","xenobiotic metabolism"],[29,29,21,12,21,15,18,10,10,9,8,9,9,8,20,9,7,7,8,7,7,9,8,9,29,6,7,6,7,7,9,6,6,9,7,10,6,6,38,5,5,5,6,5,5,5,5,5,12,4,4,4,4,6,6,4,4,7,31,4,4,4,4,4,3],[48,51,28,12,43,21,35,11,12,11,9,13,14,11,77,17,10,10,14,11,11,20,16,21,180,9,14,10,15,15,26,11,11,28,17,36,13,13,313,10,10,10,16,11,11,12,14,14,70,9,9,9,9,21,21,10,10,31,287,11,11,12,12,12,7],[1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016,1016],[15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708,15708],[9.34079724409449,8.79133858267716,11.5954724409449,15.4606299212598,7.55054019410364,11.0433070866142,7.9511811023622,14.0551181102362,12.8838582677165,12.6496062992126,13.742782152231,10.7035130224107,9.93897637795276,11.244094488189,4.01574803149606,8.18503937007874,10.8224409448819,10.8224409448819,8.83464566929134,9.83858267716535,9.83858267716535,6.95728346456693,7.73031496062992,6.6259842519685,2.49087926509186,10.3070866141732,7.73031496062992,9.2763779527559,7.21496062992126,7.21496062992126,5.35175651120533,8.43307086614173,8.43307086614173,4.96948818897638,6.36614173228346,4.29461942257218,7.13567534827377,7.13567534827377,1.87700938341174,7.73031496062992,7.73031496062992,7.73031496062992,5.79773622047244,7.02755905511811,7.02755905511811,6.44192913385827,5.52165354330709,5.52165354330709,2.6503937007874,6.87139107611548,6.87139107611548,6.87139107611548,6.87139107611548,4.41732283464567,4.41732283464567,6.18425196850394,6.18425196850394,3.49110998221996,1.66996351065873,5.62204724409449,5.62204724409449,5.15354330708661,5.15354330708661,5.15354330708661,6.6259842519685],[2.70275708666099e-20,1.51379127500772e-19,7.71184070867161e-18,4.36310849314871e-13,1.61278241032043e-12,2.84601175853215e-12,2.65928725760635e-11,5.50791523397385e-10,2.76649489011239e-09,3.23835166614882e-08,7.96826615064661e-08,3.11538932643257e-07,7.58829804021219e-07,1.02066713301989e-06,1.35086292021856e-06,6.2674517665935e-06,9.00837798664669e-06,9.00837798664669e-06,1.14813418428583e-05,2.00432555038106e-05,2.00432555038106e-05,2.63727823810346e-05,3.61615210767219e-05,3.9879882971118e-05,6.34393248229001e-05,6.81097789025051e-05,0.000136398054553117,0.00014942533653655,0.000217312367529551,0.000217312367529551,0.000244423671657209,0.000263621536119862,0.000263621536119862,0.000437967711796292,0.000501884568079371,0.000637024682522706,0.000759729415595105,0.000759729415595105,0.0011854618969614,0.00176901326236097,0.00176901326236097,0.00176901326236097,0.00264744178815494,0.00286587619907775,0.00286587619907775,0.00455076383188368,0.00988649211046072,0.00988649211046072,0.0106084487037368,0.0106084487037368,0.0106084487037368,0.0106084487037368,0.0106084487037368,0.0106084487037368,0.0106084487037368,0.0161801365613882,0.0161801365613882,0.0184589177616955,0.020542736663541,0.0225505571619588,0.0225505571619588,0.0306048189404048,0.0306048189404048,0.0306048189404048,0.0412719794864343]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>ID<\/th>\n      <th>Title<\/th>\n      <th>b<\/th>\n      <th>B<\/th>\n      <th>n<\/th>\n      <th>N<\/th>\n      <th>E<\/th>\n      <th>adj.P.Val<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"targets":8,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[3,4,5,6,7,8]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render"],"jsHooks":[]}</script>





---

## Disadvantages of the generation 1 approach

 * You have to define DEGs... but that depends on the number of samples

    * the larger the number of samples, the higher statistical power
    * the higher statistical power, the more DEGs you can detect
    * therefore, the results depend on an arbitrary threshold

 * What if you have no DEGs at all?


---
class:empty-slide,myinverse
background-image:url(images/img21.png)

---

## Generation 2 approaches – "Functional Class Sorting"

 * do not depend on arbitrary thresholds

---

## tmod algorithm

 * Arrange the genes according to p-value or another similar metrics
 * Test whether genes from a given gene set are denser on the beginning of
   the list

---
class:empty-slide,myinverse
background-image:url(images/img22.png)

---
class:empty-slide,myinverse
background-image:url(images/img23.png)

---
class:empty-slide,myinverse
background-image:url(images/img24.png)

---
class:empty-slide,myinverse
background-image:url(images/img25.png)

---
class:empty-slide,myinverse
background-image:url(images/img26.png)

---
class:empty-slide,myinverse
background-image:url(images/img27.png)

---
class:empty-slide,myinverse
background-image:url(images/img28.png)

---

### Tmod statistics

Since normalized ranking for a gene `\(i\)`

`\(r'_{i}=\frac{r_i}{N} \sim U(0,1)\)`

--

then

`\(-\ln{r'_{i}} \sim Exp(1)\)`

--

and

`\(-2\cdot\ln{r'_i} \sim \chi^2_2\)`

--

thus, if `\(G\)` is a set of genes,

`\(-2\cdot\sum_{i \in G}\ln{r'_i} \sim \sum_{i \in G}\chi^2_2 \sim \chi^2_{2\cdot n(G)}\)`

This gives a fast and reliable statistics with a known distribution which
can be used to associate a p-value to any set of genes.


---

![:scale 90%](images/tmod.png)

.footnote[*Zyla J, Marczyk M, Domaszewska T, Kaufmann SH, Polanska J, Weiner
3rd J. Gene set enrichment for reproducible science: comparison of CERNO
and eight other algorithms. Bioinformatics. 2019 Dec 15;35(24):5146-54.*]

---

![:scale 90%](images/tmod_2.png)

---

![:scale 90%](images/tmod_3.png)

.pull-left[Vaccine with an adjuvant]

.pull-right[Vaccine without an adjuvant]

.footnote[*Weiner, January, et al. "Characterization of potential biomarkers
of reactogenicity of licensed antiviral vaccines: randomized controlled
clinical trials conducted by the BIOVACSAFE consortium." Scientific reports
9.1 (2019): 1-14.*]

---

## Results in our data set

<div id="htmlwidget-575fd5a7fde04ef02fac" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-575fd5a7fde04ef02fac">{"x":{"filter":"none","data":[["LI.M75","LI.M127","LI.M165","LI.M150","LI.M7.0","LI.M67","LI.M111.1","LI.M7.3","LI.M7.2","LI.M7.1","LI.M68","LI.M13","LI.M27.1","LI.M27.0","LI.M35.1","LI.M86.1","LI.M111.0","LI.M5.0","LI.M35.0","LI.M66","LI.M86.0","LI.M38","LI.M121","LI.M36","LI.M5.1","LI.M43.1","LI.M45","LI.M11.0","LI.M115","LI.M146","LI.S11","LI.M119","LI.M29","LI.M112.0","LI.M64","LI.M52","LI.M112.1","LI.M61.2","LI.M209","LI.M19","LI.S0","LI.M21","LI.M61.0","LI.M18","LI.M44","LI.S1","LI.M12","LI.M62.0","LI.M93","LI.M126","LI.M95.0","LI.M7.4","LI.M73"],["LI.M75","LI.M127","LI.M165","LI.M150","LI.M7.0","LI.M67","LI.M111.1","LI.M7.3","LI.M7.2","LI.M7.1","LI.M68","LI.M13","LI.M27.1","LI.M27.0","LI.M35.1","LI.M86.1","LI.M111.0","LI.M5.0","LI.M35.0","LI.M66","LI.M86.0","LI.M38","LI.M121","LI.M36","LI.M5.1","LI.M43.1","LI.M45","LI.M11.0","LI.M115","LI.M146","LI.S11","LI.M119","LI.M29","LI.M112.0","LI.M64","LI.M52","LI.M112.1","LI.M61.2","LI.M209","LI.M19","LI.S0","LI.M21","LI.M61.0","LI.M18","LI.M44","LI.S1","LI.M12","LI.M62.0","LI.M93","LI.M126","LI.M95.0","LI.M7.4","LI.M73"],["antiviral IFN signature","type I interferon response","enriched in activated dendritic cells (II)","innate antiviral response","enriched in T cells (I)","activated dendritic cells","viral sensing &amp; immunity; IRF2 targets network (II)","T cell activation (II)","enriched in NK cells (I)","T cell activation (I)","RIG-1 like receptor signaling","innate activation by cytosolic DNA sensing","chemokine cluster (II)","chemokine cluster (I)","signaling in T cells (II)","proinflammatory dendritic cell, myeloid cell response","viral sensing &amp; immunity; IRF2 targets network (I)","regulation of antigen presentation and immune response","signaling in T cells (I)","TBA","chemokines and inflammatory molecules in myeloid cells","chemokines and receptors","TBA","T cell surface, activation","T cell activation and signaling","myeloid, dendritic cell activation via NFkB (II)","leukocyte activation and migration","enriched in monocytes (II)","cytokines - recepters cluster","MHC-TLR7-TLR8 cluster","Activated (LPS) dendritic cell surface signature","enriched in activated dendritic cells (I)","proinflammatory cytokines and chemokines","complement activation (I)","enriched in activated dendritic cells/monocytes","T cell activation (IV)","complement activation (II)","enriched in NK cells (receptor activation)","lysosome","T cell differentiation (Th2)","T cell surface signature","cell adhesion (lymphocyte homing)","enriched in NK cells (II)","T cell differentiation via ITK and PKC","T cell signaling and costimulation","NK cell surface signature","CD28 costimulation","T &amp; B cell development, activation","TBA","double positive thymocytes","enriched in antigen presentation (II)","T cell activation (III)","enriched in monocytes (III)"],[21,12,35,12,51,11,11,28,43,48,9,11,13,17,11,14,17,77,14,16,20,10,9,11,21,14,11,180,10,16,36,11,10,16,15,12,9,9,10,15,26,9,11,11,10,32,10,31,10,13,21,14,12],[0.844599865827634,0.994712028542304,0.803532918303543,0.904503270472307,0.824236982268158,0.966050258590234,0.979880347721336,0.88442055393586,0.784867761785643,0.815011707109408,0.963479627152473,0.878992511597468,0.842816183497929,0.742688765009541,0.854436574446767,0.85272351581132,0.767232621172872,0.674690736938834,0.711878060769357,0.755882774662248,0.752186384497705,0.886565167537266,0.871173677021183,0.855774409701912,0.804141737016092,0.781618088805549,0.888206779523592,0.590850220390406,0.854650273920245,0.758611075707367,0.623041446883331,0.799446333115187,0.827697795897567,0.678434871271986,0.732037638862338,0.794321270812096,0.744237071009477,0.888152819358629,0.675665689896802,0.726268612332462,0.605324085428664,0.831362224062396,0.70181910845732,0.775637498769307,0.727589501847369,0.585594619163052,0.686303987769143,0.604365960406348,0.741565804561091,0.651986178841865,0.703181584994551,0.751406360938666,0.731922145769623],[1.93584639118701e-19,3.65660386282376e-18,3.65660386282376e-18,1.55534843738933e-14,1.55534843738933e-14,1.09474430786983e-12,6.72831465177491e-12,6.9387176649157e-12,6.9387176649157e-12,2.10819855349691e-11,1.04244638554738e-10,3.20627727307708e-07,3.47351795035686e-07,1.34162267071206e-06,3.84617889358655e-06,4.61767628915883e-06,4.67338024009048e-06,9.57665042451423e-06,5.9380066885211e-05,6.94548250725544e-05,7.76655220113129e-05,7.76655220113129e-05,0.000103642272812651,0.000250294569984479,0.000250294569984479,0.000321746988616209,0.000451814316646038,0.000451814316646038,0.00114733020796749,0.00125504448249099,0.00129944628858933,0.00204426309201298,0.00210588798774113,0.00326961139229627,0.00381876320046766,0.00401737773597711,0.00590412028967873,0.00936921175597718,0.00936921175597718,0.0101145818896452,0.0102107726277397,0.0102552689022043,0.0140708669503597,0.0187537838755285,0.0190172450365088,0.0228079601290248,0.0375440764825181,0.0414798027362279,0.047568104192944,0.0491744667029022,0.0491844564645841,0.0497467774688221,0.0497467774688221]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>ID<\/th>\n      <th>Title<\/th>\n      <th>N1<\/th>\n      <th>AUC<\/th>\n      <th>adj.P.Val<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":8,"columnDefs":[{"targets":5,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 2, 3, \",\", \".\");\n  }"},{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[8,10,25,50,100]}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render"],"jsHooks":[]}</script>

---

## Comparison with the ORA results

![](weiner_gsea_04_files/figure-html/unnamed-chunk-10-1.svg)&lt;!-- --&gt;



---

## Effect sizes in tmod

.pull-left[

Leukocyte activation and migration: 11 genes, strongly enriched; `\(AUC = 0.89\)`, `\(p &lt; 10^{-3}\)`.

![](weiner_gsea_04_files/figure-html/unnamed-chunk-11-1.svg)&lt;!-- --&gt;

]

.pull-right[

Regulation of antigen presentation and immune response: 77 genes, weakly
enriched; `\(AUC = 0.67\)`, `\(p &lt; 10^{-5}\)`.

![](weiner_gsea_04_files/figure-html/unnamed-chunk-12-1.svg)&lt;!-- --&gt;

]


---

## GSEA algorithm

.pull-left[
![](images/GSEA-homegraphic.gif)

]

.pull-right[

**In short:**

 * Kolmogorov-Smirnof-like statistic
 * No analytical solution:
   * Good: does not make assumption of independence
   * Bad: requires randomization
   * Ugly: many implementations resample the genes rather than samples
     `\(\rightarrow\)` assumption of independence again (including original
     algorithm for sample sizes &lt; 15)
 * Really dumb name

]


---

![](images/tmod_gsea.png)

Different sample sizes were compared to the original output of each
program. *Left*, Spearman correlation of the p-values from the respective result
lists. Right, AUC (recovered gene sets compared to the original output).
GSEA stability breaks down under 15 samples, where gene-based randomization
is used.

.myfootnote[
*Weiner 3rd, January, and Teresa Domaszewska. "tmod: an R package for general and multivariate enrichment analysis." PeerJ Preprints 4 (2016).*
]

---

## Other algorithms

 * PLAGE: for each gene set, calculate a "module eigengene" – 1st component
   of a PCA. Use this instead of expression to test differences between
   groups.
 * PADOG: weight the genes by how "specific" they are for a gene set, i.e.
   genes present in multiple gene sets matter less than genes present in a
   single gene set. Also a randomization based method.
 * geneSetTest and wilcoxonGST (limma): based on gene-wise permutation.

---

![](images/comparison.jpeg)

.myfootnote[
*Zyla, Joanna, et al. "Gene set enrichment for reproducible science: comparison of CERNO and eight other algorithms." Bioinformatics 35.24 (2019): 5146-5154.*

]


---

## Network analysis (generation 3)

Genes / proteins may be linked by

 * co-expression 
 * common pathways
 * literature (co-citation)
 * synteny (co-occurence in genomes)
 * common motifs

---

## STRING

![](images/string.png)


---




## DISCO: *Discordance* / *concordance* analysis

**Concordant genes:** genes which behave similarly in a comparison.

**Discordant genes:** genes that behave differently in a comparison.

.pull-left[

**Concordant:**

![](weiner_gsea_04_files/figure-html/unnamed-chunk-14-1.svg)&lt;!-- --&gt;

]

--

.pull-right[

**Discordant:**

![](weiner_gsea_04_files/figure-html/unnamed-chunk-15-1.svg)&lt;!-- --&gt;



]

---

## DISCO statistics

The idea: include *both* effect size and p-value in the calculation of the
statistic. The sign should reflect the concordance: 
 * positive values for concordant genes, 
 * negative values for discordant genes.

 `$$\log_2(FC_1) \cdot \log_2(FC_2) \cdot -(\log_{10}(p_1) + \log_{10}(p_2))$$`


---

## DISCO plots

.center[

![](weiner_gsea_04_files/figure-html/unnamed-chunk-16-1.svg)&lt;!-- --&gt;

]

---

### DISCO + tmod

![](weiner_gsea_04_files/figure-html/unnamed-chunk-17-1.svg)&lt;!-- --&gt;

---

![](weiner_gsea_04_files/figure-html/unnamed-chunk-18-1.svg)&lt;!-- --&gt;



---

## Concordant gene sets







---

## Discordant gene sets




    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="files/macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
