<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Lecture 3: Reproducibility crisis</title>
    <meta charset="utf-8" />
    <meta name="author" content="January Weiner" />
    <meta name="date" content="2021-07-08" />
    <script src="libs/header-attrs-2.9/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="files/cubi-fonts.css" type="text/css" />
    <link rel="stylesheet" href="files/style.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Lecture 3: Reproducibility crisis
## Functional analyses in transcriptomics
### January Weiner
### 2021-07-08

---









---

## Reproducibility crisis in science

![](images/crisis_citations.png)

.myfootnote[
*Fanelli D. Opinion: Is science really facing a reproducibility crisis, and
do we need it to?. Proceedings of the National Academy of Sciences. 2018
Mar 13;115(11):2628-31.*
]

---

## Reproducibility crisis in science

![:scale 50%](images/replication_psychology.jpg)

.myfootnote[
*Open Science Collaboration. "Estimating the reproducibility of
psychological science." Science 349.6251 (2015).*
]

---

## Reproducibility crisis in science

![:scale 70%](images/Mice2.jpg)

.myfootnote[
*Perrin, Steve. "Preclinical research: Make mouse studies work." Nature
News 507.7493 (2014): 423.*
]

---

## Reproducibility crisis in science

![:scale 70%](images/ioannides.png)

.myfootnote[
*Ioannidis JP. Why most published research findings are false. PLoS medicine. 2005 Aug 30;2(8):e124.*
]

---


## Reasons for the reproducibility crisis: Money

.center[
![:scale 50%](images/research_grants.jpg)
]

---

## Reasons for the reproducibility crisis: Hype

![](images/cureforcancer.jpeg)


---

## Reproducibility crisis in science

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Journal impact factor &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Number of articles &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Mean number of citations of non reproduced articles  &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Mean number of citations of reproduced articles &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; &amp;gt;20 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 21 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 248 (range 3–800) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 231 (range 82–519) &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 5–19 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 32 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 169 (range 6–1,909) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 13 (range 3–24) &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

Bottom line: non-reproducible papers are more highly cited!

.myfootnote[
*Begley CG, Ellis LM. Raise standards for preclinical cancer research.
Nature. 2012 Mar;483(7391):531-3.*
]

---

## Reasons for the reproducibility crisis:&lt;br&gt;Misunderstanding and misusing of statistics

.center[

![](images/statistics.png)

]

---

## Reasons for the reproducibility crisis:&lt;br&gt;Misunderstanding and misusing of statistics

--

 * file drawer (suppresing non-significant results)

--

 * HARKing: Hypothesis After Results are Known

--

 * p-hacking / fishing / "researchers degrees of freedom"

--

 * "Garden of forking paths" (Gelman)


---

## Misunderstanding the p-values

What is a p-value?


---

## Cherry-picking and p-hacking 

---
class:empty-slide,myinverse
background-image:url(images/torturedata.jpg)

???

Stanisław Lem: "From strawberries under torture one may extract all sorts of
things."

---
class:empty-slide,myinverse
background-image:url(images/vonneumann.jpg)

---

## Music makes people literally younger

"*[W]e asked 20 University of Pennsylvania undergraduates to listen to
either “When I’m Sixty-Four” by The Beatles or “Kalimba.” Then, in an
ostensibly unrelated task, they indicated their birth date (mm/dd/yyyy) and
their father’s age. We used father’s age to control for variation in
baseline age across participants*.

*An ANCOVA revealed the predicted effect: According to their birth dates,
**people were nearly a year-and-a-half younger after listening to “When I’m
Sixty-Four” (adjusted M = 20.1 years) rather than to “Kalimba” (adjusted M
= 21.5 years), F(1, 17) = 4.92, p = .040**.*"

.myfootnote[ 
*Simmons JP, Nelson LD, Simonsohn U. False-positive
psychology: Undisclosed flexibility in data collection and analysis allows
presenting anything as significant. Psychological science. 2011
Nov;22(11):1359-66.* 
]


---



## A short guide to p-hacking

 * p &gt; 0.05? Try another test
 * Or maybe add or remove some covariates?
 * Surely, there are outliers. Remove them.
 * Well, let us collect some more data. And then some more. As long as it
   takes.


---

## Evidence for p-hacking: the p-curve

.pull-left[
![](images/pval_th_distribution.png)
]

.myfootnote[
left: *Head ML, Holman L, Lanfear R, Kahn AT, Jennions MD. The extent and
consequences of p-hacking in science. PLoS Biol. 2015 Mar
13;13(3):e1002106.*

right: *Masicampo EJ, Lalande DR. A peculiar prevalence of p values just below 0.05. 
Quarterly journal of experimental psychology. 2012 Nov;65(11):2271-9.*

]


--

.pull-right[
![](images/pval_distribution.png)
]


---

## Reasons for the reproducibility crisis: Lack of transparcency

.pull-left[
![](images/miyakawa.webp)
]

.pull-right[
"*We requested just the images of western blotting membranes and I wonder
how such images can contain novel information that is worth being kept
confidential.*"
]


.myfootnote[
*Miyakawa, Tsuyoshi. "No raw data, no science: another possible source of
the reproducibility crisis." (2020). Molecular Brain 13:24.*
]

---

## Reasons for the reproducibility crisis: Lack of transparcency

 * no precise description of methods:

```
We used generalized linear models (R package nlme4) to test the
significance of association.
```

Fine, but how exactly was the model formulated? How many covariates? What
transformations? What data were used? Which function? Were the results
corrected for multiple testing? Etc., etc., etc.

---

## Lack of transparency leads to different results

![](images/effect_sizes.png)

---

## p = 0.05

 * Arbitrary threshold introduced (mostly) by R.A. Fisher
 * In perfect circumstances, guarantees a false positive rate of 1 in 20

```
Q: Why do so many colleges and grad schools teach p = 0.05?

A: Because that's still what the scientific community and journal editors use.

Q: Why do so many people still use p = 0.05?

A: Because that's what they were taught in college or grad school.
```

*Wasserstein, Ronald L., and Nicole A. Lazar. "The ASA's statement on p-values: context, process, and purpose." The American Statistician (2016).*

---

## Is p = 0.05 enough?

Consider a test for a disease (like a test for Sars-Cov-2 infection). 

 * Sensitivity 80%: that is, no more than 20% False Negatives (FNs)
 * Specificity 95%: no more than 5% False Positives (FPs)

Imagine you are tested and that the result is positive. What is the
probability that you are infected with Sars-Cov-2?


(**Note:** in reality, the Sars-Cov-2 tests are *much* more sensitive and
specific)

---

## Is p = 0.05 enough?

![:scale 70%](images/cropped_hyptest_1.png)

---

## Is p = 0.05 enough?

![:scale 70%](images/cropped_hyptest_2.png)

---

## Is p = 0.05 enough?

![:scale 70%](images/cropped_hyptest_3.png)

---

## Is p = 0.05 enough?

OK, now replace "Sars-Cov-2 test" with "hypothesis test".

 * 80% sensitivity is normal for statistical tests (80% power)

 * 95% specificity corresponds to our p = 0.05

How many (scientific) hypotheses we are testing are really correct?

 * Surely much less than 100%, if it was close to 100% we would not be using
   statistics!
 * But more than 0%, we would not be doing science otherwise
 * Say, it is 10% (maybe 30% on a good day, but maybe 1% on a bad one).

---

## Is p = 0.05 enough?

![:scale 70%](images/cropped_hyptest_4.png)

---

.center[
![](weiner_reproducibility_crisis_03_files/figure-html/unnamed-chunk-4-1.png)&lt;!-- --&gt;
]

---

.center[
![](weiner_reproducibility_crisis_03_files/figure-html/unnamed-chunk-5-1.png)&lt;!-- --&gt;
]


---

## To test or not to test

The conundrum with testing for normality:

 * normality test works in the opposite direction of a regular test: 

    * rejection is the "non-favored" outcome

    * lack of significance "proofs" normality

--

 * if test is underpowered, the data will always appear normal

 * if test is overpowered, the data will *never* appear normal 

 * better: consider your prior (do you have reasons to believe there might
   be an issue with normality?) and use visualization tools / other similar
   data sets



---

.pull-left[
&lt;q&gt;To make the preliminary test on variances [before running a t-test or
ANOVA] is rather like putting to sea in a rowing boat to find out whether
conditions are sufficiently calm for an ocean liner to leave port!&lt;/a&gt;

&lt;span style="text-align:right"&gt;– George E. P. Box&lt;/span&gt;

]

.pull-right[

![:scale 80%](images/george_box.jpg)

]


---

## Multiple testing

 * Every time you run a statistical test, you have a chance of fooling
   yourself
 
 * In bioinformatics, running tens of thousands of statistical tests is
   pretty common

---
class:empty-slide,myinverse
background-image:url(images/salmon_1.png)

---

## Multiple testing

.pull-left[
**Subject.** One mature Atlantic Salmon (Salmo salar) participated in the fMRI study.
The salmon was approximately 18 inches long, weighed 3.8 lbs, and was not alive at
the time of scanning.

**Task.** The task administered to the salmon involved completing an open-ended
mentalizing task. The salmon was shown a series of photographs depicting human
individuals in social situations with a specified emotional valence. The salmon was
asked to determine what emotion the individual in the photo must have been
experiencing.
]

.pull-right[
![](images/salmon_2.png)

]

---

## Of mice and men

![](images/pnas_1.png)

---

.mycenter[

Repeatability,&lt;br&gt;
Replicability,&lt;br&gt;
Reproducibility

]

---

## Of mice and men

.pull-left[
![](images/pnas_res_1.png)
]

.pull-right[

 * No or little correlation between transcriptomic responses in mouse
   models and human models: `\(r^2 &lt; 0.1\)` – less than 10% of variance
   explained

 * No similarity in identified functions
]
 


---

## Of mice and men

![](images/pnas_1.png)

--

![](images/pnas_2.png)

---

## Of mice and men

.pull-left[
![](images/pnas_res_2.png)
]

.pull-right[

 * Strong correlation between mouse and human transcriptomic responses for
   genes which are regulated in one or the other condition:
   `\(\rho \geq 0.5\)`; change direction correlated

 * Similar functions for similarly regulated genes
]
 


---

## Reproducibility, Repeatability, Replicability


 * **Repeatability:** you can repeat your calculations precisely

.myfootnote[
*McArthur SL. Repeatability, Reproducibility, and Replicability: Tackling
the 3R challenge in biointerface science and engineering.*
]
 
--

 * **Replicability:** Others can repeat your calculations precisely.

--

 * **Reproducibility:** Other approaches to the same problem give
   compatible results.

---

## Of mice and men

.pull-left[
![](images/pnas_1.png)

![](images/pnas_2.png)
]

--

.pull-right[

 * None of the papers was replicable, because precise description of
   methods leading from data to results was missing. 


 * At best, we were able to *roughly* reproduce the results using guesswork
   and custom code


 * How can we hope for *reproducibile* research if we cannot assure
   *replicable* research?

]

---

## Of mice and men (our story)

Hypothesis: "Genomic responses in mouse models *partly* mimic human
inflammatory diseases"

---

## Of mice and men (our story)

![](images/pnas_3.png)

.myfootnote[
*Domaszewska T, Scheuermann L, Hahnke K, Mollenkopf H, Dorhoi A, Kaufmann
SH, Weiner J. Concordant and discordant gene expression patterns in mouse
strains identify best-fit animal model for human tuberculosis. Scientific
reports. 2017 Sep 21;7(1):1-3.*
]

---

## Of mice and men (our story)

.pull-left[

![](images/domaszewska.webp)
]

.pull-right[
  * numbers in the figure show numbers of concordant / discordant gene
    groups
  * 129S2 mice die when infected with *Mycobacterium tuberculosis*. Their
    responses grow more similar to human responses in the course of
    infection and less similar to C57BL/6 mice.
  * C57BL/6 mice survive the infections. The dissimilarity to human
    responses and 129S2 responses decreases over time.
]

---

## Ensuring replicability

Fist circle of hell: providing detailed information on how to replicate
your results.

.pull-left[
"*...I thought it necessary to deliver things circumstantially, that the
person I addressed them to might, without mistake, and with as little
trouble as possible, be able to repeat such unusual experiments*"

*Robert Boyle, 1627-1691*
]

.pull-right[
![:scale 70%](images/boyle.jpg)
]


---
class:empty-slide,myinverse
background-image:url(images/data_analysis_scheme.png)


---
class:empty-slide,myinverse
background-image:url(images/data_analysis_scheme_2.png)

---

## Ensuring replicability

 * publish your code

--

 * make sure that the code *starts* with data exactly as it is available
   for your colleagues (no custom Excel files)

--

 * avoid at all costs any operations that can not be written down as code
   (e.g. editing of spreadsheets in Excel *unless* you want to publish the
   edited versions)

--

 * The code should produce output which is as close to the published
   results (figures, tables etc.) as possible.

--

 * Provide generous instructions on how to run your code, which software or
   R packages need to be installed etc. Make it as simple as possible;
   optimally, it should run without problems with a single simple command.

---

## Ensuring replicability

Second circle of hell: providing detailed information on the software you
have used.



---

## Ensuring replicability: `sessionInfo()`

**Always** include `sessionInfo()` at the end of your report.


```r
sessionInfo()
```

---

&lt;pre class="r-output"&gt;&lt;code&gt;## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=de_DE.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] cowplot_1.1.1    forcats_0.5.1    stringr_1.4.0    dplyr_1.0.7      purrr_0.3.4      readr_1.4.0     
##  [7] tidyr_1.1.3      tibble_3.1.2     ggplot2_3.3.5    tidyverse_1.3.1  crayon_1.4.1     kableExtra_1.3.4
## [13] knitr_1.33       pander_0.6.4    
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.6         svglite_2.0.0      lubridate_1.7.10   ps_1.6.0           assertthat_0.2.1   digest_0.6.27     
##  [7] utf8_1.2.1         R6_2.5.0           cellranger_1.1.0   backports_1.2.1    reprex_2.0.0       evaluate_0.14     
## [13] httr_1.4.2         xaringan_0.22      highr_0.9          pillar_1.6.1       rlang_0.4.11       readxl_1.3.1      
## [19] rstudioapi_0.13    jquerylib_0.1.4    rmarkdown_2.9      labeling_0.4.2     webshot_0.5.2      munsell_0.5.0     
## [25] broom_0.7.8        compiler_4.1.0     modelr_0.1.8       xfun_0.24          pkgconfig_2.0.3    systemfonts_1.0.2 
## [31] htmltools_0.5.1.1  tidyselect_1.1.1   fansi_0.5.0        viridisLite_0.4.0  dbplyr_2.1.1       withr_2.4.2       
## [37] grid_4.1.0         jsonlite_1.7.2     gtable_0.3.0       lifecycle_1.0.0    DBI_1.1.1          magrittr_2.0.1    
## [43] scales_1.1.1       cli_2.5.0          stringi_1.6.2      farver_2.1.0       fs_1.5.0           xml2_1.3.2        
## [49] bslib_0.2.5.1      ellipsis_0.3.2     generics_0.1.0     vctrs_0.3.8        RColorBrewer_1.1-2 tools_4.1.0       
## [55] glue_1.4.2         hms_1.1.0          yaml_2.2.1         colorspace_2.0-2   rvest_1.0.0        haven_2.4.1       
## [61] sass_0.4.0
&lt;/code&gt;&lt;/pre&gt;

---

## Ensuring replicability: other software

 * If possible, use an automated system to include the software version you
   use

--

The following code runs the NCBI program `blastp` with the option
`-version` and captures the output of the program in the variable
`blastp_version`.


```r
## running in R, so we can include it in Rmarkdown
blastp_version &lt;- system2("blastp", "-version", stdout=TRUE)
blastp_version &lt;- blastp_version[2]
```

--

Then, one can use it directly in Rmarkdown:

```
To search for similar amino acid sequences, we have used the BLASTP program
included in the BLAST package ( ` r blastp_version ` ).
```



--

Output:

To search for similar amino acid sequences, we have used the BLASTP program
included in the BLAST package ( Package: blast 2.9.0, build Sep 30 2019 01:57:31).

---

## Ensuring replicability

Further five circles of hell: actually making sure that your software will
run.

 * Package versions behave differently and produce different output
 * Even different versions of R may produce different output

--

Solution: 

 * Use a VM / Container / Environment to freeze the state of the software
   (program versions, options etc.)
 * Major downside: takes up a lot of space
 

---

## Ensuring replicability: VMs


Virtual Machines: all software including OS in a single file which can be
moved, archived and "executed". It will freeze the state of every single
component of your analysis pipeline. 

--

![:scale 10%](images/vbox_logo.png)

E.g. Oracle VM [VirtualBox](https://www.virtualbox.org/) – free, available for Windows, MacOS and Linux,
can run any guest-OS on any host-OS (e.g. Windows on Linux, or Linux on
MacOS etc.).

Windows Substem Linux (WSL) on Windows.

--

.pull-left[
Advantages:

 * portable
 * easy to use
 * ensures 100% software identity
]

.pull-right[
Downsides: 

 * huge performance penalty 
 * images may include proprietary software (e.g. Windows OS), and often cannot be freely published.
]

---
class:empty-slide,myinverse
background-image:url(images/vbox.png)

---

## Ensuring replicability: Containers

Containers: very similar to VM, but do not actually run a virtual
machine. Instead, they depend on the host OS to run your code, but all
software / libraries are contained in the container.

![:scale 10%](images/singularity_logo.png)

E.g. [Singularity](https://sylabs.io/guides/3.0/user-guide/index.html) (runs on Linux only)

.pull-left[
Advantages:

 * no performance hit
 * portable
 * efficient
 * have many more applications (ensuring portability)
]

.pull-right[
Downsides:

 * much harder to use
 * command-line only
 * basically Linux (running on Windows / MacOS using a VM with Linux)
]

---

## Ensuring replicability: Environments

Environments: lightweight approach to containers. Not all software is
really contained, and the integration (e.g. choosing which software version
to run) happens over environmental variables. 

![](images/anaconda_logo.png)

E.g. [Anaconda](https://www.anaconda.com/): Python-based, works on Linux, MacOS, Windows. Software
packages can be specified with a simple text file, versions can be
selected. The environment is stored in a user directory tree.

--

.pull-left[
Advantages:
 
 * relatively easier to use
 * repeatability: +
 * can be used to describe the installed environment in much detail but in
   a simple text file

]

.pull-right[
Downsides:

 * portability-, replicability- (compared to containers / VMs)
 * package dependency hell
 * package availability might be limited
]



---

## Reproducible bioinformatic analyses with R

.pull-left[
![](images/Rlogo.svg)
]

.pull-right[

 * Writing reproducible and communicative analyses

 * Using markdown for your everyday needs
 
 * Tools for transparent and reproducible research (git, github, pipelines, containers)

 * Get a jump start on R – which makes reproducible research much easier
   and is a great statistical analysis language

]


---

## Why learn reproducible science with R?

 * Easy to document your statistical procedures

--

 * Harder to hide harking, p-hacking and file drawer

--

 * One way or the other, you will have to learn it


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="files/macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
