<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Lecture 1: Introduction</title>
    <meta charset="utf-8" />
    <meta name="author" content="January Weiner" />
    <meta name="date" content="2021-06-19" />
    <script src="libs/header-attrs-2.9/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="files/cubi-fonts.css" type="text/css" />
    <link rel="stylesheet" href="files/style.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Lecture 1: Introduction
## Functional analyses in transcriptomics
### January Weiner
### 2021-06-19

---









## Berlin Institute of Health

.pull-left[

 * Institute for translational and precision medicine

 * Founded in 2013 by Max Delbrück Center for
   Molecular Medicine and Charité – a university hospital affiliated with
   both Humboldt University and Free University Berlin.

 * Since 2020 a part of the Charité

]


![](images/charite.jpg)


???

Charité – over 15,000 employes


---

## A few words about myself

.pull-left[

Short bio:

 * PhD in molecular biology 
 * 2003-2009 evolutionary bioinformatics (sequence analyses, phylogeny, teaching)
 * 2009-2019 Max Planck for Infection Biology – tuberculosis, Omics, statistics, clinical trials
 * Since 2019 at the Core Unit for Bioinformatics, BIH, Charité

Areas of expertise:

 * biostatistics, R, "bioinformatics"¹
 * functional analysis
 * evolutionary biology
]

.myfootnote[
¹ Molecular biologists were so stunned at the idea of using statistics and
computers in research that they invented a new word for it.
]

--

.pull-right[

![:scale 60%](images/technika.jpg)

]

---

## A few words about myself

.pull-left[

Short bio:

 * PhD in molecular biology 
 * 2003-2009 evolutionary bioinformatics (sequence analyses, phylogeny, teaching)
 * 2009-2019 Max Planck for Infection Biology – tuberculosis, Omics, statistics, clinical trials
 * Since 2019 at the Core Unit for Bioinformatics, BIH, Charité

Areas of expertise:

 * biostatistics, R, "bioinformatics"¹
 * functional analysis
 * evolutionary biology
]

.myfootnote[
¹ Molecular biologists were so stunned at the idea of using statistics and
computers in research that they invented a new word for it.
]

.pull-right[

![:scale 60%](images/biokompost.png)

[Biokompost](https://biokompost.wordpress.com)

]

---

## Organisatory matters

 * Course materials repository: [https://github.com/january3/FunctionalTranscriptomics](https://github.com/january3/FunctionalTranscriptomics)

--

 * Direct access to lectures (HTML): [https://january3.github.io/FunctionalTranscriptomics/](https://january3.github.io/FunctionalTranscriptomics/)

--

 * Getting credit for this course: you will have to send me a reproducible
   analysis (Rmarkdown or Jupyter notebook or a similar system) and the
   corresponding analysis files (not PDF or HTML!). I will have to be able
   to run it / view it. You will receive the homework during the second
   Workshop session and you will have two weeks to complete it. 

--

It won't be hard.


--

 * Please ask questions.

--

 * **Please do *ask* questions.**

---

## What this course is about

 * Functional analyses in transcriptomics
 * Reproducibility in science
    * Using R to take care of a *small bit* of a larger issue
 * Jump-starting on R/Rstudio


---

## Reproducibility crisis in science

![](images/crisis_citations.png)

.myfootnote[
*Fanelli D. Opinion: Is science really facing a reproducibility crisis, and
do we need it to?. Proceedings of the National Academy of Sciences. 2018
Mar 13;115(11):2628-31.*
]

---

## Reproducibility crisis in science

![:scale 50%](images/replication_psychology.jpg)

.myfootnote[
*Open Science Collaboration. "Estimating the reproducibility of
psychological science." Science 349.6251 (2015).*
]

---

## Reproducibility crisis in science

![:scale 70%](images/Mice2.jpg)

.myfootnote[
*Perrin, Steve. "Preclinical research: Make mouse studies work." Nature
News 507.7493 (2014): 423.*
]

---

## Reproducibility crisis in science

![:scale 70%](images/ioannides.png)

.myfootnote[
*Ioannidis JP. Why most published research findings are false. PLoS medicine. 2005 Aug 30;2(8):e124.*
]

---

## Reasons for the reproducibility crisis: Lack of transparcency

.pull-left[
![](images/miyakawa.webp)
]

.pull-right[
"*We requested just the images of western blotting membranes and I wonder
how such images can contain novel information that is worth being kept
confidential.*"
]


.myfootnote[
*Miyakawa, Tsuyoshi. "No raw data, no science: another possible source of
the reproducibility crisis." (2020). Molecular Brain 13:24.*
]

---

## Reasons for the reproducibility crisis: Lack of transparcency

 * no precise description of methods:

```
We used generalized linear models (R package nlme4) to test the
significance of association.
```

Fine, but how exactly was the model formulated? How many covariates? What
transformations? What data were used? Which function? Were the results
corrected for multiple testing? Etc., etc., etc.

---

## What you will learn here and how that will help

.pull-left[
![](images/Rlogo.svg)
]

.pull-right[

 * Functional analysis of transcriptomic data

 * Writing reproducible and communicative analyses

 * Using markdown for your everyday needs
 
 * Tools for transparent and reproducible research (git, github, pipelines, containers)

 * Get a jump start on R – which makes reproducible research much easier
   and is a great statistical analysis language

]


---

## Why learn reproducible science with R?

 * Easy to document your statistical procedures

--

 * Fast lane from data to publication

--

 * Even if you barely use R, Rstudio is a great interface for editing
   (R)markdown files and github.

--

 * Harder to hide harking, p-hacking and file drawer

--

 * You can use it with Python, too

--

 * One way or the other, you will have to learn it or something similar

---

## Alternative: Jupyter Notebook

 * Python-oriented, but you can also use it with R

 * I find it harder to set up and use

 * No git interface

 * Not really an IDE (integrated development environment)

 * Less statistical analysis oriented, more flexible

---

## Why learn R?

.pull-left[

#### Pro

 * Massive amount of tools for statistics and specifically bioinformatics

 * Easy to get help

 * Even if you use Python, you will have to learn R (or you will work much
   more efficiently if you know R)

 * Nice if you like functional languages

 * R shiny allows creating server-based interactive apps

 * Learning statistics is *way* harder than learning R


]

--

.pull-right[

#### Con

 * Messy, bottom-up development with many quirks and inconsistencies

 * Many "dialects" (base R, tidyverse, Bioconductor...)

 * Functional languages are not intuitive if you come from a different
   programming background

]

---

## Python vs R

.pull-left[


### Python

 * flexible
 * general programming language
 * deep learning algorithms
 * you can use R from Python

]

.pull-right[

### R
 
 * statistics
 * data visualizations
 * interactive data exploration
 * low overhead for quick tasks
 * you can use Python from R

]

--

**You probably will have to learn both, at least the basics.**

---

## Course plan

 * Workshop session 1: setting up our working environment
   
    * Lecture: Introduction

    * Introduction to R / Rstudio

    * Introduction to (R)markdown

    * Introduction to git / github

    * Very basic crash-course like introduction to R

    * Introduction to high-throughput data analysis, transcriptomics

    * Running a basic data analysis in transcriptomics

    * Lecture: Introduction to the problems of functional analysis

---

## Course plan


 * Workshop session 2: functional analysis of transcriptomic data

    * Introduction to clustering a.k.a. unsupervised machine learning

    * Clustering genes 

    * Introduction to gene set enrichments

    * Gene set enrichment analysis


---

## Course plan

 * Lecture: Statistical analyses in times of reproducibility crisis

 * Lecture: Gene set enrichment algorithms – an overview

---

## Rstudio introduction

 * Open Rstudio
 * Open the workshop 1/1 HTML file
 * We go through it together


---


.pull-left[

#### The R language

 * Interpreted language
 * Functional language: everything is a function
 * Vectorization
 * Pass by value 
 * Lazy evaluation
 * Object oriented
 * Four main types of objects:
    * Vectors
    * Matrices / Arrays
    * Lists
    * Functions

]

--

.pull-right[

#### Functional language features:

 * Minimization of side effects
 * Closures (anonymous functions, like lambda in Python)
 * Vectorization: 
   * operating on vectors rather than scalars
   * Apply / map instead of loops

]

---

## Introduction to R

 * Return to the workshop 1/1 file.
 * Proceed until the end of the "R language basis" section.


---

## Literate programming


.pull-left[

*I believe that the time is ripe for significantly better documentation
of programs, and that we can best achieve this by considering
programs to be works of literature. Hence, my title: “Literate
Programming.”*

*Let us change our traditional attitude to the construction of
programs: Instead of imagining that our main task is to instruct a
computer what to do, let us concentrate rather on explaining to
human beings what we want a computer to do.*

(Donald E. Knuth)
]

.pull-right[
![Donald E. Knuth](images/knuth.jpg)
]

---

## Coding for others (including yourself)

 * the "future you" will need to understand your analysis. You think you will
   remember, but you won't.

--

 * Principles of reproducibility require that your analysis is readable by
   other humans, not only by machines

--

 * Clear, well documented and communicative analysis will be also easier to
   maintain in future

---
 
## Literate programming

*The practitioner of literate programming can be regarded as an
essayist, whose main concern is with exposition and excellence of
style. Such an author, with thesaurus in hand, **chooses the names of
variables carefully and explains what each variable means**.*

--

*He or she strives for a program that is comprehensible because its concepts
have been introduced in **an order that is best for human understanding**,
using a mixture of formal and informal methods that reinforce each other.*

(Donald E. Knuth)


---

## Introduction to (R)markdown

 * Proceed with the workshop 1/1 file until the end.





    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script src="files/macros.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"countIncrementalSlides": false,
"ratio": "16:9"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
